### DISCUSSION TITLE
Transaction query already complete (Should not run flush in parallel ?)

### DISCUSSION THREAD

**Discussion by dungkaka** [mikro-orm/mikro-orm#5372] [General]
In my education system, when student Enroll a class, I also want to make an orderHistory for this action.
So, let see below code (just demonstrate)

```ts
addEnrollments = (enrollmentUpserts) => {
      await this.em.upsertMany(Enroll, enrollmentUpserts, {
            onConflictFields: ["user", "course"],
            onConflictAction: "ignore"
        });
        return true;
}

createOrderHistories = (orderHistories) => {
       const newOrderHistories = orderHistories.map((oH) => this.create(oH));
        await this.em.persistAndFlush(newOrderHistories);
        return true;

}

await Promise.all([addEnrollments(), createOrderHistories()]
]
```

And sometime (just very few), I throw error: 
"DriverException: Transaction query already complete, run with DEBUG=knex:tx for more info"

Like mention in #1885, https://github.com/mikro-orm/mikro-orm/issues/1885#issuecomment-851583093:

> You should not use single EM instance in parallel.

I wonder if my case is the same as the above case? And then I should not use Promise.all() to run flush in parallel ? 

MikroORM: 6.0.5
NodeJS: 20

**Comment by B4nan** [comment]
Nowadays flushing in paralle is supported, but it in fact does not do anything in parallel, the calls are synchronized as the UoW has it a queue for the flush calls. Your problem is the combination with `upsert`, if this would be two flush calls, it would work fine most probably.

One way around those problems is using separate fork for each "process":

```ts
addEnrollments = (enrollmentUpserts) => {
      await this.em.fork().upsertMany(Enroll, enrollmentUpserts, {
            onConflictFields: ["user", "course"],
            onConflictAction: "ignore"
        });
        return true;
}

createOrderHistories = (orderHistories) => {
       const newOrderHistories = orderHistories.map((oH) => this.create(oH, { persist: false }));
        await this.em.fork().persistAndFlush(newOrderHistories);
        return true;

}

await Promise.all([addEnrollments(), createOrderHistories()]
```

note the `em.fork()` calls, also i disabled the automatic persist from `em.create` as you want to persist them in the fork which is created later

**Reply by dungkaka** [reply]
Thank u, but I also wonder why the problem is "combination with upsert" ? I guess that upsert is not UoW then it may be consider like "parallel" ?

**Reply by B4nan** [reply]
the flush call will create a transaction context, the upsert call will try to use it, but it will be closed at some point (when the flush resolves), i guess that's what is happening.

**Reply by dungkaka** [reply]
Thank you so much, now I understood !

**Reply by leonardbinet** [reply]
👋  I'm having a similar issue, where the stack-trace points to code where I perform a `find` query (not a mutation):

```
DriverException: Transaction query already complete, run with DEBUG=knex:tx for more info
    at PostgreSqlExceptionConverter.convertException (/home/node/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8:16)
    at PostgreSqlExceptionConverter.convertException (/home/node/node_modules/@mikro-orm/postgresql/PostgreSqlExceptionConverter.js:48:22)
    at PostgreSqlDriver.convertException (/home/node/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:351:54)
    at /home/node/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:355:24
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at PostgreSqlDriver.find (/home/node/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:71:24)
    at SqlEntityManager.find (/home/node/node_modules/@mikro-orm/core/EntityManager.js:145:25)
    at SqlEntityManager.find (/home/node/node_modules/@mikro-orm/core/EntityManager.js:114:25)
    at SQLMappingRowRepository.list (/home/node/src/utils/mikroOrm/repository/base.repository.ts:119:20)
    at RemoveAutogeneratedDuplicateMappingRowsService.removeAutogeneratedDuplicates (/home/node/src/datacollection/domain/mappingTable/service/mappingTable/remove.autogenerated.duplicate.mappingRows.service.ts:26:18)

    at completedError (/home/node/node_modules/knex/lib/execution/transaction.js:408:9)
    at /home/node/node_modules/knex/lib/execution/transaction.js:374:24
    at new Promise (<anonymous>)
    at PostgreSqlKnexDialect.trxClient.query (/home/node/node_modules/knex/lib/execution/transaction.js:370:12)
    at Runner.query (/home/node/node_modules/knex/lib/execution/runner.js:141:36)
    at ensureConnectionCallback (/home/node/node_modules/knex/lib/execution/internal/ensure-connection-callback.js:13:17)
    at Runner.ensureConnection (/home/node/node_modules/knex/lib/execution/runner.js:318:20)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Runner.run (/home/node/node_modules/knex/lib/execution/runner.js:30:19)
```
the concerned code is a `await this.repository.find`, I wouldn't have suspected "find" queries to be affected. (note: called with `disableIdentityMap: true`, for perf reasons in this very specific case and `refresh: true`)

**Reply by B4nan** [reply]
https://mikro-orm.io/docs/unit-of-work#flush-modes

**Reply by B4nan** [reply]
Hmm, but this does not look like an auto flush, more like you run an explicit transaction which gets closed from elsewhere.

**Reply by leonardbinet** [reply]
It might seem a bit dumb, but to be honest, I struggle to understand in which circumstances this error can happen, and what is the expected behaviour when multiple concurrent queries are run with the same entity manager.

In initially provided example:
`the flush call will create a transaction context, the upsert call will try to use it, but it will be closed at some point (when the flush resolves), i guess that's what is happening.`

what does it mean that `upsert call will try to use it` ? is the expected behaviour that both statements are run within the same transaction?
- if so, that's quite unexpected from my point of view (we might want to have those 2 queries succeeding or failing independently)
- if not, why do they have to share a transaction context

**Reply by B4nan** [reply]
If the `em.flush` opens a transaction, and the `em.upsert` call gets executed afterwards, it will reuse this transaction, since the EM will have a valid tx context, but it might get closed too early. It was just a guess, I am also not entirely sure this was the problem, but it sounds like it. Flushing does open a tx [here](https://github.com/mikro-orm/mikro-orm/blob/83ca9c14cc4fd443b45ea40f1945c5a249336747/packages/core/src/unit-of-work/UnitOfWork.ts#L414) and [here](https://github.com/mikro-orm/mikro-orm/blob/83ca9c14cc4fd443b45ea40f1945c5a249336747/packages/core/src/unit-of-work/UnitOfWork.ts#L932) we set the tx context to the EM itself. So if this happens first, then `em.upsert` will respect it.

> if so, that's quite unexpected from my point of view (we might want to have those 2 queries succeeding or failing independently)

Then don't run them concurrently, or isolate their contexts (use separate forks as suggested above). The EM itself is what holds the transaction context.

**Reply by leonardbinet** [reply]
`Then don't run them concurrently, or isolate their contexts (use separate forks as suggested above). The EM itself is what holds the transaction context.` 🤦 indeed, my brain cells just connected.

`more like you run an explicit transaction which gets closed from elsewhere.` I've triple checked, it is not the case, we don't use explicit transactions in this case.

FYI, I end-up digging this topic because we encountered this kind of issue `Transaction query already complete` in two contexts:
- in prod, very rarely (above example)
- in CI

For the CI, we have e2e-tests which:
- run an instance of our web app (nestjs)
- another process triggers some "black-box" e2e tests, via jest, that call our endpoints.

Jest configuration allowed multiple processes (simultaneous tests).

While bumping from mikro 6.4.15 to 6.5.4, some tests started to fail with this error, in a deterministic way.
 - 6.4.15 (multiple // jest workers): ✅ 
 - 6.5.4 (multiple // jest workers) : 🔴  (deterministic)
 - 6.5.4 (single jest worker) : ✅ 
By configuring jest to use a single worker, it now works flawlessly.

I have no idea if that can bring any idea. I wouldn't have necessarily expected the change of number of workers to impact the result of this test. Is there any chance that the error message isn't right, and that it's more a connection/db issue?

**Reply by B4nan** [reply]
>  I've triple checked, it is not the case, we don't use explicit transactions in this case.

Could be an implicit one (so one created via `em.flush`) too.

> Is there any chance that the error message isn't right, and that it's more a connection/db issue?

Hard to guess, it's not coming from our code, but I wouldn't expect that.

**Reply by leonardbinet** [reply]
To sum up my above example:
- a `find` query (not a mutation), fails because `Transaction query already complete`
- there is no explicit transaction usage

So, wouldn't it be a bug in this case then? given what you said about `flushing in paralle is supported, but it in fact does not do anything in parallel, the calls are synchronized as the UoW has it a queue for the flush calls.`


Note: in CI, I've run the CI with `DEBUG=knex:tx`, and here is the result:
```
2025-09-19T11:57:51.2159226Z 2025-09-19T11:57:51.215Z knex:tx trx256: Starting top level transaction
2025-09-19T11:57:51.2219620Z 2025-09-19T11:57:51.221Z knex:tx trx256: releasing connection
2025-09-19T11:57:51.2251893Z 2025-09-19T11:57:51.224Z knex:tx trx257: Starting top level transaction
2025-09-19T11:57:51.2344945Z 2025-09-19T11:57:51.234Z knex:tx trx257: Transaction completed: select "e0".* from "emission_factor" as "e0" where "e0"."id" in (3, 4)
2025-09-19T11:57:51.2376024Z 
2025-09-19T11:57:51.2377385Z /__w/traace/traace/backend/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8
2025-09-19T11:57:51.2378341Z         return new exceptions_1.DriverException(exception);
2025-09-19T11:57:51.2378784Z                ^
2025-09-19T11:57:51.2379575Z DriverException: Transaction query already complete, run with DEBUG=knex:tx for more info
...
```

the first 255 transactions all have both: `Starting top level transaction` and `releasing connection`
but the 256'th transaction ends with a `Transaction completed: select "e0".* from "emission_factor" as "e0" where "e0"."id" in (3, 4)` 🤔 


note: I don't manage to reproduce the issue locally, it only fails in CI..

**Reply by B4nan** [reply]
> a find query (not a mutation), fails because Transaction query already complete

I believe that if you see this, it means the EM you are calling `em.find` on has a tx context. 

> So, wouldn't it be a bug in this case then?

Likely in your code, though. It still sounds like you work with a single EM concurrently, and it opens a transaction in one context, while the other context is using it for find queries only.

**Reply by leonardbinet** [reply]
There is no concurrent use of the EM in this specific use-case, but the bug might be deeper on our side: in our setup, each incoming request / bullmq job gets a fork of the EM, stored in the async-local-storage. The fact that in the CI the bug appears when there is more load on the server (because multiple jest workers are running tests in // on the same web-server) might indicate that the fork isn't properly applied (I have no idea why this wouldn't be the case, but I'm still investigating).

Thanks for your help


Note: we fork only with default values (ie `em.fork()`), I'll double check our setup, and any misuse of `ForkOptions`

**Reply by leonardbinet** [reply]
mmh, actually, there might be a bug in knex : https://github.com/knex/knex/issues/4709, last line: `hides query timeout and instead returns a completed transaction object` and indeed:
- when there's a read timeout (or connection issue) during the initial BEGIN operation, the transaction gets silently marked as completed
- instead of seeing the actual timeout/connection error, a "Transaction query already complete" is thrown

**Reply by leonardbinet** [reply]
note, I've not checked in mikro-orm how knex is used, but this fix/work-around might be applied in mikro-orm as well: https://github.com/supabase/storage/pull/459/files

**Reply by B4nan** [reply]
We indeed use `connection.transaction()` ([here](https://github.com/mikro-orm/mikro-orm/blob/083a1eec70c065c49023d10363d976662e1daf4a/packages/knex/src/AbstractSqlConnection.ts#L101)). The PR that workarounds this feels quite cryptic to me, and given we no longer use knex in v7, I'd leave things as they are now. 

But if you see an easy way to do the same in our code, feel free to PR this. We await the execution promise here, if that matters:

https://github.com/mikro-orm/mikro-orm/blob/083a1eec70c065c49023d10363d976662e1daf4a/packages/knex/src/AbstractSqlConnection.ts#L136

**Reply by leonardbinet** [reply]
it's currently not a big deal for us, I was mostly scared of having an issue in my connections/EMs management, so we'll wait for the v7 👍

**Reply by leonardbinet** [reply]
Actually, we encountered again the issue, and after deeper investigation the issue was actually an issue with our ORM usage, and not because of knex.

## Investigation
I managed to reproduce the bug locally in one of our e2e tests.

By patching knex to add more logs/details when this issue occurs, I noticed it was a legit case of "Transaction query already complete".

With debug logs, I noticed lots of empty transactions that were unintended:
<img width="773" height="421" alt="Screenshot 2025-10-10 at 10 35 24" src="https://github.com/user-attachments/assets/a25e6bb9-d182-41dc-84e0-77211ffd2f6d" />

That's where I found out about the default flush mode `auto` (which I was not aware of).

Long story short, by switching flush mode strategy from `auto`(default)  to `commit` this fixed the issue and removed the unintended empty transactions.

## Context

This happened in a graphql mutation context, with usage of resolve fields.

In our setup, in nestjs, when a new request kicks in, we fork the entity manager, and get a clean entity manager for the whole request lifecycle.

But a single request lifecycle can include:
- multiple simultaneous @Mutations or @Queries
- lots of @ResolveField usages

Which can reuse the same EM in multiple concurrent places.

That's why when calling @Mutations we only call one mutation per graphql request.

## What might have happened

Graphql request is received, targeting a single mutation.
- entity manager is forked
- the mutation succeeds
- the resolved fields are simultaneously called, using the same entity manager
   - in one/multiple resolved fields, the auto flush mechanism triggers an empty transaction
   - one of the other resolve field using the same EM ends up querying while the transaction context references a closed transaction

<img width="804" height="533" alt="Screenshot 2025-10-10 at 11 35 05" src="https://github.com/user-attachments/assets/5169f549-ee79-4370-9d36-0eb4e5e7b8aa" />


### Additional thoughts
If I didn't get it wrong, this might be worth adding a "pitfall" panel in documentation about the dangers of reusing an entity manager  in graphql context, if the request targets multiple endpoints or resolved fields.

**Reply by leonardbinet** [reply]
note, I've opened a dedicated issue for this topic https://github.com/mikro-orm/mikro-orm/issues/6916
